### PR TITLE
[IMP] mail:tests: use mail.channel.partner for seen indicator

### PR DESCRIPTION
**Pull request by alexkuhn**
Task-2810539

Enterprise: https://github.com/odoo/enterprise/pull/25911

### CODE REVIEW

File: addons/im_livechat/static/tests/qunit_suite_tests/components/chat_window_manager_tests.js [hunk 1]
```diff
@@ -18,21 +19,18 @@ QUnit.module('chat_window_manager_tests.js', {
 QUnit.test('closing a chat window with no message from admin side unpins it', async function (assert) {
```
Comment (alexkuhn) [inline addons/im_livechat/static/tests/qunit_suite_tests/components/chat_window_manager_tests.js:19]: This test has been adapted to use `pyEnv`, because change in `MockServer/_mockMailChannelChannelFetched` requires having a `mail.channel.partner` record being made automatically, and it's working with `pyEnv` but not with `this.data` .

I think it comes from `_mockCreate` not properly called with `this.data`, which itself doesn't auto-create `mail.channel.partner` from `MockServer/_mockMailChannel__inverseChannelPartnerIds`.

File: addons/mail/static/tests/helpers/mock_server.js [hunk 2]
```diff
@@ -314,6 +314,16 @@ MockServer.include({
         }
         return mockWriteResult;
     },
+    /**
+     * @override
+     */
+    _writeRecord(model, values, id) {
```
Comment (alexkuhn) [inline addons/mail/static/tests/helpers/mock_server.js:320]: Solution was to overwrite `_writeRecord` to have this behaviour in both `pyEnv` and `this.data` the safest way possible.

The adapt of `im_livechat` above could be reverted. But we want sooner or later to adapt, so up to reviewer whether it's fine or prefer revert.

**Comment by alexkuhn** [comment]
Putting this PR on hold until all discuss tests use `pyEnv`.

File: addons/mail/static/tests/helpers/mock_server.js [hunk 3]
```diff
@@ -1888,10 +1940,10 @@ MockServer.include({
 
             // notify update of last_interest_dt
             const now = datetime_to_str(new Date());
-            this._mockWrite('mail.channel',
+            this._mockWrite('mail.channel', [
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:1943]: Why this change? Wrong syntax? Shouldn't it be done as a fix?

File: addons/mail/static/tests/helpers/mock_server.js [hunk 4]
```diff
@@ -1323,10 +1350,35 @@ MockServer.include({
...
+            fetched_message_id: message_id,
+            seen_message_id: message_id,
+        }]);
+    },
+    /**
+     * Simulates `_inverse_channel_partner_ids` on `mail.channel`
+     *
+     * @private
+     */
+     _mockMailChannel__inverseChannelPartnerIds(channelIds) {
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:1369]: single _ in the middle

File: addons/mail/static/tests/helpers/mock_server.js [hunk 5]
```diff
@@ -1323,10 +1350,35 @@ MockServer.include({
...
+        for (const channelId of channelIds) {
+            const channel = this._getRecords('mail.channel', [['id', 'in', [channelId]]])[0];
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:1371]: _getRecords can work in batch, could do it outside then loop on the result

File: addons/mail/static/tests/helpers/mock_server.js [hunk 6]
```diff
@@ -1323,10 +1350,35 @@ MockServer.include({
...
+            for (const memberId of channel.channel_partner_ids) {
+                const cp = this._getRecords('mail.channel.partner', [['partner_id', '=', memberId], ['channel_id', '=', channelId]])[0];
+                if (!cp) {
+                    this._mockCreate('mail.channel.partner', {
+                        channel_id: channelId,
+                        partner_id: memberId,
+                    });
+                }
+            }
+        }
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:1381]: also missing removal of existing member that are gone

File: addons/mail/static/tests/helpers/mock_server.js [hunk 7]
```diff
@@ -1323,10 +1350,35 @@ MockServer.include({
      * @param {integer} message_id
      */
     _mockMailChannel_SetLastSeenMessage(ids, message_id) {
+        // deprecated: update mail.channel/fetched_message_id and mail.channel/seen_message_id
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:1353]: mail.channel meant to be mail.channel.partner?

File: addons/mail/static/tests/helpers/mock_server.js [hunk 8]
```diff
@@ -830,10 +848,16 @@ MockServer.include({
             if (!lastMessage) {
                 continue;
             }
+            // deprecated: update mail.channel/fetched_message_id
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:851]: mail.channel.partner?

File: addons/mail/static/tests/helpers/mock_server.js (hunk 3 above)
Comment (alexkuhn) [reply addons/mail/static/tests/helpers/mock_server.js:1943]: Just spotted wrong syntax, I think I got buggy behaviour due to this typo.

Comment (seb-odoo) [reply addons/mail/static/tests/helpers/mock_server.js:1943]: I think the diff is correct, but it should be done as fix outside of this PR?

2nd param of _mockWrite is supposed to be an array of args as you did.

File: addons/mail/static/tests/helpers/mock_server.js (hunk 4 above)
Comment (alexkuhn) [reply addons/mail/static/tests/helpers/mock_server.js:1369]: The method in python is `_inverse_channel_partner_ids`, so there's an underscore before.
There is no convention to not use similar name regardless of private/public prefix with `_`, so I prefer to consider `_f` and `f` as 2 different names, to play it safe.

Comment (seb-odoo) [reply addons/mail/static/tests/helpers/mock_server.js:1369]: Yes but non _ method have 0 not 1 (it's just camel case).

File: addons/mail/static/tests/helpers/mock_server.js (hunk 7 above)
Comment (alexkuhn) [reply addons/mail/static/tests/helpers/mock_server.js:1353]: It intentionally refers to `mail.channel` (fake) fields.

File: addons/mail/static/tests/helpers/mock_server.js (hunk 8 above)
Comment (alexkuhn) [reply addons/mail/static/tests/helpers/mock_server.js:851]: same remark, it's refers to a fake field on `mail.channel`.
This is just backward-compatibility for behaviour that still rely on this fake field.

This is to not force ourselves to adapt them right now, which could slow down the making of this PR, which is a requirement for many other PRs.

File: addons/mail/static/tests/helpers/mock_server.js (hunk 4 above)
Comment (alexkuhn) [reply addons/mail/static/tests/helpers/mock_server.js:1369]: Oh ok, I thought there was at least one `_` to separate `modelName` and `methodName`.

File: addons/mail/static/tests/helpers/mock_server.js [hunk 9]
```diff
@@ -711,6 +721,8 @@ MockServer.include({
                 channel_partner_ids: [[3, this.currentPartnerId]],
             },
         ]);
+        const cp = this._getRecords('mail.channel.partner', [['partner_id', '=', this.currentPartnerId], ['channel_id', '=', channel.id]])[0];
+        this._mockUnlink('mail.channel.partner', [[cp.id]]);
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:725]: As commented elsewhere this should be done in `_mockMailChannel_inverseChannelPartnerIds` when records are removed. Here there should be no code change.

File: addons/mail/static/tests/helpers/mock_server.js (hunk 6 above)
Comment (alexkuhn) [reply addons/mail/static/tests/helpers/mock_server.js]: I don't think we should manage deletion here. Server does it but it's quite obscure how it manages `channel_last_seen_partner_ids`. There is no such handling of this field in the mock server.

If we had some code for `_action_remove_members`, then we would simply remove `mail.channel.partner` record during this action.

File: addons/mail/static/tests/helpers/mock_server.js (hunk 8 above)
Comment (seb-odoo) [reply addons/mail/static/tests/helpers/mock_server.js:851]: Ok, wording is a bit confusing, usually with deprecated notice we tell what should be used instead, rather than what is wrong. This looks more like a todo.

File: addons/mail/static/tests/helpers/mock_server.js (hunk 3 above)
Comment (alexkuhn) [reply addons/mail/static/tests/helpers/mock_server.js:1943]: Made a 15.0 PR that fix this `_mockWrite` call: https://github.com/odoo/odoo/pull/88009

File: addons/mail/static/tests/helpers/mock_server.js (hunk 8 above)
Comment (seb-odoo) [reply addons/mail/static/tests/helpers/mock_server.js:851]: Not sure if the comment is worth it then.

File: addons/mail/static/tests/helpers/mock_server.js [hunk 10]
```diff
@@ -830,10 +848,16 @@ MockServer.include({
             if (!lastMessage) {
                 continue;
             }
+            // deprecated: update mail.channel/fetched_message_id
             this._mockWrite('mail.channel', [
                 [channel.id],
                 { fetched_message_id: lastMessage.id },
             ]);
+            const [cp] = this._getRecords('mail.channel.partner', [['channel_id', 'in', [channel.id]], ['partner_id', '=', this.currentPartnerId]]);
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:856]: mix of operator "in" and "=", both are correct but lack of consistency is strange

File: addons/mail/static/tests/helpers/mock_server.js [hunk 11]
```diff
@@ -975,11 +999,14 @@ MockServer.include({
...
-                    fetched_message_id: channel.fetched_message_id,
-                }];
+                res['seen_partners_info'] = res.channel_partner_ids.map(memberId => {
+                    const [cp] = this._getRecords('mail.channel.partner', [['channel_id', 'in', [channel.id]], ['partner_id', '=', memberId]]);
+                    return {
+                        partner_id: memberId,
+                        seen_message_id: cp.seen_message_id,
+                        fetched_message_id: cp.fetched_message_id,
+                    };
+                });
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:1009]: If the ONLY source of truth for this value is now the channel partner and not the channel, I think all occurrences in tests/mock server of fake field should be removed in this PR. It's not keeping the compatibility of the old syntax now does nothing.

File: addons/mail/static/tests/helpers/mock_server.js (hunk 8 above)
Comment (alexkuhn) [reply addons/mail/static/tests/helpers/mock_server.js:850]: Just added comment below to state what we should use.
Let me know if you prefer this or just remove comment (or another suggestion for comment, or something else?)

File: addons/mail/static/tests/helpers/mock_server.js (hunk 6 above)
Comment (seb-odoo) [reply addons/mail/static/tests/helpers/mock_server.js]: It's important to do it, it should be declarative (members in the list = actual members), not flux specific.

Just need to find all channel.partner for this channel, and delete those that are not in channel_partner_ids.

Note that for consistency it requires the opposite works too, if that is not yet the case: when creating a channel.partner, it should fill the related channel_partner_ids, and remove from itwhen deleting.

File: addons/mail/static/tests/helpers/mock_server.js (hunk 11 above)
Comment (alexkuhn) [reply addons/mail/static/tests/helpers/mock_server.js:1009]: Yeah, I did not want to necessarily remove it, but I think it's best just to drop old fake fields altogether.

File: addons/mail/static/tests/helpers/mock_server.js (hunk 7 above)
Comment (alexkuhn) [reply addons/mail/static/tests/helpers/mock_server.js]: In the end, I just removed the `mail.channel` fake fields `fetched_message_id` and `seen_message_id`.

File: addons/mail/static/tests/helpers/mock_server.js [hunk 12]
```diff
@@ -1034,7 +1053,8 @@ MockServer.include({
         if (!channel) {
             return;
         }
-        if (channel.seen_message_id && channel.seen_message_id >= last_message_id) {
+        const cp = this._getRecords('mail.partner.channel', [['id', '=', channel_id]])[0];
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:1056]: should take for current partner id only

File: addons/mail/static/tests/helpers/mock_server.js [hunk 13]
```diff
@@ -1323,11 +1343,37 @@ MockServer.include({
...
+            let outdatedCps = this._getRecords('mail.channel.partner', [['channel_id', '=', channel.id]]);
+            for (const memberId of channel.channel_partner_ids) {
+                const cp = this._getRecords('mail.channel.partner', [['partner_id', '=', memberId], ['channel_id', '=', channel.id]])[0];
+                if (!cp) {
+                    this._mockCreate('mail.channel.partner', {
+                        channel_id: channel.id,
+                        partner_id: memberId,
+                    });
+                } else {
+                    outdatedCps = outdatedCps.filter(cp => cp.partner_id !== memberId);
```
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:1369]: could have domain for `outdatedCps` to be `partner_id not in channel.channel_partner_ids` instead?

**Comment by seb-odoo** [comment]
red runbot

**Comment by alexkuhn** [comment]
TODO: debug test "sidebar: pinned channel 3: open channel and leave it"

Somehow there's a `_mockWrite` that has deep code in `_updateComodelRelationalFields` makes a problematic `channel_id` read on `undefined`.

**Comment by seb-odoo** [comment]
need to update channel_partner_ids when creating or deleting mail.channel.partner

**Comment by alexkuhn** [comment]
@robodoo r+ rebase-ff
